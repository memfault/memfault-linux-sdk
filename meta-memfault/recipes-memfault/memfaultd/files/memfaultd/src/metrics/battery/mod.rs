//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod battery_reading_handler;
pub use battery_reading_handler::BatteryReadingHandler;

mod battery_monitor;
pub use battery_monitor::BatteryMonitor;
pub use battery_monitor::BatteryMonitorReading;
