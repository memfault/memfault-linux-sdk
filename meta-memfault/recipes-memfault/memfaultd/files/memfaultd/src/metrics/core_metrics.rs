//
// Copyright (c) Memfault, Inc.
// See License.txt for details
// Connectivity metrics
pub const METRIC_MF_SYNC_SUCCESS: &str = "sync_memfault_successful";
pub const METRIC_MF_SYNC_FAILURE: &str = "sync_memfault_failure";
pub const METRIC_CONNECTED_TIME: &str = "connectivity_connected_time_ms";
pub const METRIC_EXPECTED_CONNECTED_TIME: &str = "connectivity_expected_time_ms";
pub const METRIC_SYNC_SUCCESS: &str = "sync_successful";
pub const METRIC_SYNC_FAILURE: &str = "sync_failure";

// Stability metrics
pub const METRIC_OPERATIONAL_HOURS: &str = "operational_hours";
pub const METRIC_OPERATIONAL_CRASHFREE_HOURS: &str = "operational_crashfree_hours";
pub const METRIC_OPERATIONAL_CRASHES: &str = "operational_crashes";

// Battery metrics
pub const METRIC_BATTERY_DISCHARGE_DURATION_MS: &str = "battery_discharge_duration_ms";
pub const METRIC_BATTERY_SOC_PCT_DROP: &str = "battery_soc_pct_drop";
pub const METRIC_BATTERY_SOC_PCT: &str = "battery_soc_pct";
