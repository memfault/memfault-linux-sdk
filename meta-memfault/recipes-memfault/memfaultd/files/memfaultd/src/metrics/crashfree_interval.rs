//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    cmp::max,
    collections::HashMap,
    io::Read,
    iter::once,
    str::{from_utf8, FromStr},
    sync::mpsc::{channel, Receiver, Sender},
    time::Duration,
};

use eyre::{eyre, Result};
use log::{trace, warn};
use serde::{Deserialize, Serialize};
use tiny_http::{Method, Request, Response};

use crate::{
    cli::NotifyCrashRequest,
    http_server::{HttpHandler, HttpHandlerResult},
    metrics::{
        core_metrics::{
            METRIC_OPERATIONAL_CRASHES, METRIC_OPERATIONAL_CRASHES_PROCESS_PREFIX,
            METRIC_OPERATIONAL_CRASHFREE_HOURS, METRIC_OPERATIONAL_HOURS,
        },
        KeyedMetricReading,
    },
    util::time_measure::TimeMeasure,
};

use super::{MetricStringKey, MetricsMBox};

pub struct CrashFreeIntervalTracker<T: TimeMeasure> {
    last_interval_mark: T,
    last_crashfree_interval_mark: T,
    crash_count: u32,
    process_crash_count: HashMap<String, u64>,
    sender: Sender<CrashInfo<T>>,
    receiver: Receiver<CrashInfo<T>>,
    interval: Duration,
    elapsed_intervals_key: MetricStringKey,
    crashfree_intervals_key: MetricStringKey,
    crash_count_key: MetricStringKey,
    metrics_mbox: MetricsMBox,
}

#[derive(Debug, PartialEq, Eq)]
struct TimeMod<T: TimeMeasure> {
    count: u32,
    mark: T,
}

impl<T> CrashFreeIntervalTracker<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync + 'static,
{
    pub fn new(
        interval: Duration,
        elapsed_intervals_key: &'static str,
        crashfree_intervals_key: &'static str,
        crash_count_key: &'static str,
        metrics_mbox: MetricsMBox,
    ) -> Self {
        let (sender, receiver) = channel();
        Self {
            last_crashfree_interval_mark: T::now(),
            last_interval_mark: T::now(),
            sender,
            receiver,
            crash_count: 0,
            process_crash_count: HashMap::new(),
            interval,
            elapsed_intervals_key: MetricStringKey::from(elapsed_intervals_key),
            crashfree_intervals_key: MetricStringKey::from(crashfree_intervals_key),
            crash_count_key: MetricStringKey::from(crash_count_key),
            metrics_mbox,
        }
    }

    /// Returns a tracker with an hourly interval
    pub fn new_hourly(metrics_mbox: MetricsMBox) -> Self {
        Self::new(
            Duration::from_secs(3600),
            METRIC_OPERATIONAL_HOURS,
            METRIC_OPERATIONAL_CRASHFREE_HOURS,
            METRIC_OPERATIONAL_CRASHES,
            metrics_mbox,
        )
    }

    /// Wait for the next crash or update the metrics if the wait duration has passed.
    ///
    /// This allows us to have instant updates on crashes and hourly updates on the metrics, but
    /// also allows us to periodically update the metrics so that we don't have to wait for a crash.
    pub fn wait_and_update(&mut self, wait_duration: Duration) -> Result<()> {
        if let Ok(crash_info) = self.receiver.recv_timeout(wait_duration) {
            // Drain the receiver to get all crashes that happened since the last update
            self.receiver
                .try_iter()
                .chain(once(crash_info))
                .for_each(|info| {
                    self.crash_count += 1;
                    self.last_crashfree_interval_mark =
                        max(self.last_crashfree_interval_mark, info.timestamp);
                    *self
                        .process_crash_count
                        .entry(info.process_name)
                        .or_insert(0) += 1;
                });
        }

        // Since timing out just means no crashes occurred in the `wait_duration`,
        // update even when the receiver times out.
        self.update()
    }

    fn update(&mut self) -> Result<()> {
        let TimeMod {
            count: count_op_interval,
            mark: last_counted_op_interval,
        } = Self::full_interval_elapsed_since(self.interval, &self.last_interval_mark);
        let TimeMod {
            count: count_crashfree_interval,
            mark: last_counted_crashfree_interval,
        } = Self::full_interval_elapsed_since(self.interval, &self.last_crashfree_interval_mark);

        self.last_interval_mark = last_counted_op_interval;
        self.last_crashfree_interval_mark = last_counted_crashfree_interval;

        let crashes = self.crash_count;
        self.crash_count = 0;

        let mut metrics = vec![
            KeyedMetricReading::add_to_counter(
                self.elapsed_intervals_key.clone(),
                count_op_interval as f64,
            ),
            KeyedMetricReading::add_to_counter(
                self.crashfree_intervals_key.clone(),
                count_crashfree_interval as f64,
            ),
            KeyedMetricReading::add_to_counter(self.crash_count_key.clone(), crashes as f64),
        ];

        let process_crash_metrics = self.process_crash_count.drain().flat_map(
            |(process_name, crash_count)| -> Result<KeyedMetricReading> {
                let operation_crashes_process_key = MetricStringKey::from_str(
                    format!(
                        "{}{}",
                        METRIC_OPERATIONAL_CRASHES_PROCESS_PREFIX, process_name
                    )
                    .as_str(),
                )
                .map_err(|e| eyre!("Couldn't construct MetricStringKey: {}", e))?;
                Ok(KeyedMetricReading::add_to_counter(
                    operation_crashes_process_key,
                    crash_count as f64,
                ))
            },
        );

        metrics.extend(process_crash_metrics);

        trace!("Crashfree hours metrics: {:?}", metrics);

        self.metrics_mbox.send_and_forget(metrics)?;

        Ok(())
    }

    pub fn http_handler(&mut self) -> Box<dyn HttpHandler> {
        Box::new(CrashFreeIntervalHttpHandler {
            channel: self.sender.clone(),
        })
    }

    pub fn capture_crash(&self, process_name: String) {
        self.sender
            .send(CrashInfo {
                process_name,
                timestamp: T::now(),
            })
            .expect("Failed to send crash timestamp");
    }

    /// Count how many `interval` have elapsed since `since`.
    ///
    /// This returns the number of intervals that have elapsed since `since`, and the timestamp of the end of the last interval
    /// that was counted. This is the value you should pass as `since` next time you call this function.
    ///
    /// See unit test for examples.
    fn full_interval_elapsed_since(interval: Duration, since: &T) -> TimeMod<T> {
        let now = T::now();
        if *since > now {
            return TimeMod {
                count: 0,
                mark: T::now(),
            };
        }

        let duration = now.since(since);
        let count_interval_elapsed = (duration.as_nanos() / interval.as_nanos()) as u32;
        TimeMod {
            count: count_interval_elapsed,
            mark: since.add(interval * count_interval_elapsed),
        }
    }
}

struct CrashFreeIntervalHttpHandler<T> {
    channel: Sender<CrashInfo<T>>,
}

#[derive(Serialize, Deserialize)]
struct CrashInfo<T> {
    process_name: String,
    timestamp: T,
}

impl<T> CrashFreeIntervalHttpHandler<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync,
{
    fn parse_notify_crash_request(stream: &mut dyn Read) -> Result<NotifyCrashRequest> {
        let mut buf = vec![];
        stream.read_to_end(&mut buf)?;
        let reading = serde_json::from_str(from_utf8(&buf)?)?;
        Ok(reading)
    }
}

impl<T> HttpHandler for CrashFreeIntervalHttpHandler<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync,
{
    fn handle_request(&self, request: &mut Request) -> HttpHandlerResult {
        if request.url() == "/v1/crash/report" && request.method() == &Method::Post {
            match Self::parse_notify_crash_request(request.as_reader()) {
                Ok(NotifyCrashRequest { process_name }) => {
                    self.channel
                        .send(CrashInfo {
                            process_name,
                            timestamp: T::now(),
                        })
                        .expect("Crashfree channel closed");
                }
                Err(e) => {
                    warn!("Failed to parse notify crash request: {}", e);
                    return HttpHandlerResult::Error(format!(
                        "Failed to notify crash request: {}",
                        e
                    ));
                }
            }
            HttpHandlerResult::Response(Response::from_string("OK").boxed())
        } else {
            HttpHandlerResult::NotHandled
        }
    }
}

#[cfg(test)]
mod tests {
    use std::{collections::BTreeMap, time::Duration};

    use insta::assert_json_snapshot;
    use rstest::rstest;
    use ssf::ServiceMock;

    use crate::{
        metrics::{
            crashfree_interval::{
                METRIC_OPERATIONAL_CRASHES, METRIC_OPERATIONAL_CRASHFREE_HOURS,
                METRIC_OPERATIONAL_HOURS,
            },
            MetricStringKey, MetricValue, TakeMetrics,
        },
        test_utils::TestInstant,
    };

    use super::CrashFreeIntervalTracker;
    use super::TimeMod;

    #[rstest]
    fn test_counting_intervals() {
        use std::time::Duration;

        // move the clock forward so we can go backwards below
        TestInstant::sleep(Duration::from_secs(3600));
        let now = TestInstant::now();

        let d10 = Duration::from_secs(10);
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(d10, &now),
            TimeMod {
                count: 0,
                mark: now
            }
        );
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(
                d10,
                &(now - Duration::from_secs(10))
            ),
            TimeMod {
                count: 1,
                mark: now
            }
        );
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(
                d10,
                &(now - Duration::from_secs(25))
            ),
            TimeMod {
                count: 2,
                mark: now - Duration::from_secs(5)
            }
        );
    }

    #[rstest]
    fn test_counting_hours() {
        let mut metrics_mock = ServiceMock::new();
        let mut crashfree_tracker =
            CrashFreeIntervalTracker::<TestInstant>::new_hourly(metrics_mock.mbox.clone());

        TestInstant::sleep(Duration::from_secs(7200));

        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();

        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 2, 2);
    }

    #[rstest]
    fn test_counting_minutes() {
        let mut metrics_mock = ServiceMock::new();
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new(
            Duration::from_secs(60),
            METRIC_OPERATIONAL_HOURS,
            METRIC_OPERATIONAL_CRASHFREE_HOURS,
            METRIC_OPERATIONAL_CRASHES,
            metrics_mock.mbox.clone(),
        );

        TestInstant::sleep(Duration::from_secs(3600));
        crashfree_tracker.capture_crash("memfaultd".to_string());
        TestInstant::sleep(Duration::from_secs(3600));

        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 120, 60);
    }

    #[rstest]
    fn test_30min_heartbeat() {
        let mut metrics_mock = ServiceMock::new();
        let mut crashfree_tracker =
            CrashFreeIntervalTracker::<TestInstant>::new_hourly(metrics_mock.mbox.clone());

        TestInstant::sleep(Duration::from_secs(1800));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 0, 0);

        TestInstant::sleep(Duration::from_secs(1800));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 1, 1);
    }

    #[rstest]
    fn test_30min_heartbeat_with_crash() {
        let mut metrics_mock = ServiceMock::new();
        let mut crashfree_tracker =
            CrashFreeIntervalTracker::<TestInstant>::new_hourly(metrics_mock.mbox.clone());

        TestInstant::sleep(Duration::from_secs(1800));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 0, 0);

        // Crash at t0 + 30min
        crashfree_tracker.capture_crash("memfaultd".to_string());

        // After 30' we should be ready to mark an operational hour
        TestInstant::sleep(Duration::from_secs(1800));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 1, 0);

        // After another 30' we should be ready to mark another crashfree hour
        TestInstant::sleep(Duration::from_secs(1800));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 0, 1);

        // After another 30' we should be ready to mark another operational hour
        TestInstant::sleep(Duration::from_secs(1800));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 1, 0);
    }

    #[rstest]
    fn test_180min_heartbeat_with_one_crash() {
        let mut metrics_mock = ServiceMock::new();
        let mut crashfree_tracker =
            CrashFreeIntervalTracker::<TestInstant>::new_hourly(metrics_mock.mbox.clone());

        // Basic test
        TestInstant::sleep(Duration::from_secs(3600 * 3));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 3, 3);

        // Crash at interval + 170'
        TestInstant::sleep(Duration::from_secs(170 * 60));
        crashfree_tracker.capture_crash("memfaultd".to_string());

        // Another 10' to the heartbeat mark
        // We will count 0 operational hour here. That is a consequence of the heartbeat being larger than the hour
        // To avoid this bug, we need to make sure we call the `update` at least once per hour!
        TestInstant::sleep(Duration::from_secs(10 * 60));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 3, 0);

        // However, doing the crash at interval +10' then waiting for 170' will record 2 crashfree hours
        TestInstant::sleep(Duration::from_secs(10 * 60));
        crashfree_tracker.capture_crash("memfaultd".to_string());
        TestInstant::sleep(Duration::from_secs(170 * 60));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_operational_metrics(metrics_mock.take_metrics().unwrap(), 3, 2);
    }

    #[rstest]
    fn test_process_crash_counter() {
        let mut metrics_mock = ServiceMock::new();
        let mut crashfree_tracker =
            CrashFreeIntervalTracker::<TestInstant>::new_hourly(metrics_mock.mbox.clone());

        crashfree_tracker.capture_crash("memfaultd".to_string());
        crashfree_tracker.capture_crash("memfaultd".to_string());
        crashfree_tracker.capture_crash("memfaultd".to_string());
        crashfree_tracker.capture_crash("collectd".to_string());

        TestInstant::sleep(Duration::from_secs(3600));
        crashfree_tracker
            .wait_and_update(Duration::from_secs(0))
            .unwrap();
        assert_json_snapshot!(metrics_mock.take_metrics().unwrap());
    }

    fn assert_operational_metrics(
        metrics: BTreeMap<MetricStringKey, MetricValue>,
        expected_op_hours: u32,
        expected_crashfree_hours: u32,
    ) {
        let op_hours = metrics
            .iter()
            .find(|(name, _)| name.as_str() == METRIC_OPERATIONAL_HOURS)
            .map(|(_, value)| value)
            .unwrap();
        let crash_free_hours = metrics
            .iter()
            .find(|(name, _)| name.as_str() == METRIC_OPERATIONAL_CRASHFREE_HOURS)
            .map(|(_, value)| value)
            .unwrap();

        let op_hours_value = match op_hours {
            MetricValue::Number(value) => value,
            _ => panic!("Unexpected metric type"),
        };

        let crashfree_hours_value = match crash_free_hours {
            MetricValue::Number(value) => value,
            _ => panic!("Unexpected metric type"),
        };

        assert_eq!(
            (*op_hours_value as u32, *crashfree_hours_value as u32),
            (expected_op_hours, expected_crashfree_hours)
        );
    }
}
