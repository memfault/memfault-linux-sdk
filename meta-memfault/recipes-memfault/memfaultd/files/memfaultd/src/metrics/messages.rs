//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::path::PathBuf;

use eyre::Result;
use ssf::{Message, MsgMailbox};

use crate::network::NetworkConfig;

use super::{KeyedMetricReading, SessionName};

/// Allows KeyedMetricReading to be sent as a message. The `ssf` framework will
/// automatically support sending `Vec<KeyedMetricReading>` as well.
impl Message for KeyedMetricReading {
    type Reply = Result<()>;
}

/// Syntactic Sugar because this is the most-used type of mailbox in memfaultd.
pub type MetricsMBox = MsgMailbox<Vec<KeyedMetricReading>>;

pub enum SessionEventMessage {
    StartSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
    },
    StopSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
        mar_staging_area: PathBuf,
        network_config: NetworkConfig,
    },
}

impl Message for SessionEventMessage {
    type Reply = Result<()>;
}
