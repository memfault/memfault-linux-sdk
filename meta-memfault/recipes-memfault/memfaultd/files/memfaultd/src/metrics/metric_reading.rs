//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::str::FromStr;

use chrono::{Duration, Utc};
use eyre::{eyre, ErrReport};
use nom::{
    branch::alt,
    character::complete::char,
    combinator::value,
    number::complete::double,
    sequence::separated_pair,
    Finish,
    {bytes::complete::tag, IResult},
};
use serde::{Deserialize, Serialize};

use super::{MetricStringKey, MetricTimestamp};

use crate::util::serialization::float_to_duration;

#[derive(Clone, Debug, Serialize, Deserialize)]
/// A typed value and timestamp pair that represents
/// an individual reading value for a metric. This type does
/// not have a notion of which key it is associated it
/// and is purely the "value" in a metric reading.
/// For the full metric reading type that includes the key,
/// use KeyedMetricReading.
pub enum MetricReading {
    /// TimeWeightedAverage readings will be aggregated based on
    /// the time the reading was captured over.
    TimeWeightedAverage {
        value: f64,
        timestamp: MetricTimestamp,
        /// Time period considered for this reading. This is only used to give a "time-weight" to the first
        /// value in the series. For future values we will use the time
        /// difference we measure between the two points
        /// In doubt, it's safe to use Duration::from_secs(0) here. This means the first value will be ignored.
        #[serde(with = "float_to_duration")]
        interval: Duration,
    },
    /// A non-decreasing monotonic sum. Within a metric report, Counter readings are summed together.
    Counter {
        value: f64,
        timestamp: MetricTimestamp,
    },
    /// Gauges are absolute values. We keep the latest value collected during a metric report.
    Gauge {
        value: f64,
        timestamp: MetricTimestamp,
    },
    /// Histogram readings are averaged together by dividing the sum of the values
    /// by the number of readings over the duration of a metric report.
    Histogram {
        value: f64,
        timestamp: MetricTimestamp,
    },
    /// ReportTags are string values associated with the MetricReport they are captured in.
    /// We keep the latest value collected during a metric report for a key and drop the older
    /// ones.
    ReportTag {
        value: String,
        timestamp: MetricTimestamp,
    },
}

impl MetricReading {
    /// Parse a metric reading in the format f64|<StatsDMetricType>.
    /// The timestamp will be set to Utc::now().
    /// This is the suffix of a full StatsD reading of the following format:
    /// <MetricStringKey>:<MetricReading>
    ///
    /// Examples of valid readings:
    /// 64|h
    /// 100.0|c
    /// -89.5|g
    fn parse(input: &str) -> IResult<&str, MetricReading> {
        let (remaining, (value, statsd_type)) =
            separated_pair(double, tag("|"), StatsDMetricType::parse)(input)?;
        let timestamp = Utc::now();
        match statsd_type {
            StatsDMetricType::Histogram => {
                Ok((remaining, MetricReading::Histogram { value, timestamp }))
            }
            StatsDMetricType::Counter => {
                Ok((remaining, MetricReading::Counter { value, timestamp }))
            }
            StatsDMetricType::Timer => Ok((remaining, MetricReading::Counter { value, timestamp })),
            StatsDMetricType::Gauge => Ok((remaining, MetricReading::Gauge { value, timestamp })),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct KeyedMetricReading {
    pub name: MetricStringKey,
    pub value: MetricReading,
}

impl KeyedMetricReading {
    pub fn new(name: MetricStringKey, value: MetricReading) -> Self {
        Self { name, value }
    }

    pub fn new_gauge(name: MetricStringKey, value: f64) -> Self {
        Self {
            name,
            value: MetricReading::Gauge {
                value,
                timestamp: Utc::now(),
            },
        }
    }

    pub fn new_histogram(name: MetricStringKey, value: f64) -> Self {
        Self {
            name,
            value: MetricReading::Histogram {
                value,
                timestamp: Utc::now(),
            },
        }
    }

    pub fn new_counter(name: MetricStringKey, value: f64) -> Self {
        Self {
            name,
            value: MetricReading::Counter {
                value,
                timestamp: Utc::now(),
            },
        }
    }

    /// Construct a KeyedMetricReading from a string in the StatsD format
    /// <MetricStringKey:<f64>|<StatsDMetricType>
    ///
    /// Examples of valid keyed metric readings:
    ///  testCounter:1|c
    ///  test_counter:1.0|c
    ///  test_histo:100|h
    ///  test_gauge:1.7|g
    ///  cpu3_idle:100.9898|g
    pub fn from_statsd_str(s: &str) -> Result<Self, ErrReport> {
        match Self::parse_statsd(s).finish() {
            Ok((_, reading)) => Ok(reading),
            Err(e) => Err(eyre!(
                "Failed to parse string \"{}\" as a StatsD metric reading: {}",
                s,
                e
            )),
        }
    }

    pub fn increment_counter(name: MetricStringKey) -> Self {
        Self {
            name,
            value: MetricReading::Counter {
                value: 1.0,
                timestamp: Utc::now(),
            },
        }
    }

    pub fn add_to_counter(name: MetricStringKey, value: f64) -> Self {
        Self {
            name,
            value: MetricReading::Counter {
                value,
                timestamp: Utc::now(),
            },
        }
    }

    /// Helper that handles `nom` details for parsing a StatsD string as
    /// a KeyedMetricReading
    fn parse_statsd(input: &str) -> IResult<&str, KeyedMetricReading> {
        let (remaining, (name, value)) =
            separated_pair(MetricStringKey::parse, tag(":"), MetricReading::parse)(input)?;
        Ok((remaining, KeyedMetricReading { name, value }))
    }

    /// Deserialize a string in the form <MetricStringKey>=<f64> to a Gauge metric reading
    ///
    /// Currently deserialization to a KeyedMetricReading with any other type of MetricReading
    /// as its value is not supported
    fn from_arg_str(s: &str) -> Result<Self, ErrReport> {
        let (key, value_str) = s.split_once('=').ok_or(eyre!(
            "Attached metric reading should be specified as KEY=VALUE"
        ))?;

        // Let's ensure the key is valid first:
        let metric_key = MetricStringKey::from_str(key).map_err(|e| eyre!(e))?;
        if let Ok(value) = f64::from_str(value_str) {
            let reading = MetricReading::Gauge {
                value,
                timestamp: Utc::now(),
            };
            Ok(KeyedMetricReading::new(metric_key, reading))
        } else {
            let reading = MetricReading::ReportTag {
                value: value_str.to_string(),
                timestamp: Utc::now(),
            };
            Ok(KeyedMetricReading::new(metric_key, reading))
        }
    }
}

impl FromStr for KeyedMetricReading {
    type Err = ErrReport;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match Self::from_statsd_str(s) {
            Ok(reading) => Ok(reading),
            Err(_) => match Self::from_arg_str(s) {
                Ok(reading) => Ok(reading),
                Err(e) => Err(eyre!("Couldn't parse \"{}\" as a Gauge metric: {}", s, e)),
            },
        }
    }
}

#[derive(Debug, Clone)]
enum StatsDMetricType {
    Counter,
    Histogram,
    Gauge,
    Timer,
}

impl StatsDMetricType {
    /// Parse a StatsDMetricType, which must be one of 'c', 'g', or 'h'
    ///
    /// 'c' indicates a Counter reading
    /// 'g' indicates a Gauge reading
    /// 'h' indicates a Histogram reading
    fn parse(input: &str) -> IResult<&str, StatsDMetricType> {
        alt((
            value(StatsDMetricType::Counter, char('c')),
            value(StatsDMetricType::Histogram, char('h')),
            value(StatsDMetricType::Gauge, char('g')),
            value(StatsDMetricType::Timer, tag("ms")),
        ))(input)
    }
}

#[cfg(test)]
mod tests {

    use super::*;
    use crate::test_utils::setup_logger;
    use rstest::rstest;

    #[rstest]
    #[case("testCounter=1")]
    #[case("hello=world")]
    #[case("float=100.0")]
    fn parse_valid_arg_reading(#[case] reading_str: &str, _setup_logger: ()) {
        assert!(KeyedMetricReading::from_str(reading_str).is_ok())
    }

    #[rstest]
    #[case("testCounter:1|c")]
    #[case("test_counter:1.0|c")]
    #[case("test_histo:100|h")]
    #[case("test_gauge:1.7|g")]
    #[case("cpu3_idle:100.9898|g")]
    #[case("some_negative_gauge:-87.55|g")]
    #[case("test_timer:3600000|ms")]
    fn parse_valid_statsd_reading(#[case] reading_str: &str, _setup_logger: ()) {
        assert!(KeyedMetricReading::from_str(reading_str).is_ok())
    }

    #[rstest]
    #[case("test Counter:1|c")]
    #[case("{test_counter:1.0|c}")]
    #[case("\"test_counter\":1.0|c}")]
    #[case("test_gauge:\"string-value\"|g")]
    #[case("test_gauge:string-value|g")]
    fn fail_on_invalid_statsd_reading(#[case] reading_str: &str, _setup_logger: ()) {
        assert!(KeyedMetricReading::from_str(reading_str).is_err())
    }
}
