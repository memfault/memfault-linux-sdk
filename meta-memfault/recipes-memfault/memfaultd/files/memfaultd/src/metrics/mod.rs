//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, Utc};

mod battery;
pub use battery::BatteryMonitor;
pub use battery::BatteryMonitorReading;
pub use battery::BatteryReadingHandler;

mod connectivity;
pub use connectivity::ConnectivityMonitor;
pub use connectivity::ReportSyncEventHandler;

mod metric_string_key;
pub use metric_string_key::MetricStringKey;

mod metric_report;
pub use metric_report::MetricReport;
pub use metric_report::MetricReportType;

mod metric_report_manager;
pub use metric_report_manager::MetricReportManager;
pub use metric_report_manager::TakeMetrics;

mod messages;
pub use messages::*;

mod metric_reading;
pub use metric_reading::KeyedMetricReading;
pub use metric_reading::MetricReading;

mod timeseries;

mod metric_value;
pub use metric_value::MetricValue;

pub type MetricTimestamp = DateTime<Utc>;

mod periodic_metric_report;
pub use periodic_metric_report::PeriodicMetricReportDumper;

mod crashfree_interval;
pub use crashfree_interval::CrashFreeIntervalTracker;

mod session_name;
pub use session_name::SessionName;

mod session_event_handler;
pub use session_event_handler::SessionEventHandler;

pub mod core_metrics;

pub mod statsd_server;
pub use statsd_server::StatsDServer;

mod system_metrics;
pub use system_metrics::SystemMetricsCollector;
pub use system_metrics::BUILTIN_SYSTEM_METRIC_NAMESPACES;
