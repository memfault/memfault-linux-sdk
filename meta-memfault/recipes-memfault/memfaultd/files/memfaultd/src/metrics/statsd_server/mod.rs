//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::net::SocketAddr;
use std::net::UdpSocket;

use eyre::Result;
use log::warn;

use crate::metrics::KeyedMetricReading;

use super::MetricsMBox;

pub struct StatsDServer {
    metrics_mailbox: MetricsMBox,
}

impl StatsDServer {
    pub fn new(metrics_mailbox: MetricsMBox) -> StatsDServer {
        StatsDServer { metrics_mailbox }
    }

    pub fn run(&self, listening_address: SocketAddr) -> Result<()> {
        let socket = UdpSocket::bind(listening_address)?;
        loop {
            // This means that packets with > 1432 bytes are NOT supported
            // Clients must enforce a maximum message size of 1432 bytes or less
            let mut buf = [0; 1432];
            match socket.recv(&mut buf) {
                Ok(amt) => {
                    let message = String::from_utf8_lossy(&buf[..amt]);
                    self.process_statsd_message(&message)
                }
                Err(e) => warn!("Statsd server socket error: {}", e),
            }
        }
    }

    fn process_statsd_message(&self, message: &str) {
        // https://github.com/statsd/statsd/blob/master/docs/server.md
        // From statsd spec:
        // Multiple metrics can be received in a single packet if separated by the \n character.
        let metric_readings = message
            .trim()
            .lines()
            .map(KeyedMetricReading::from_statsd_str)
            // Drop strings that couldn't be parsed as a KeyedMetricReading
            .filter_map(|res| {
                if let Err(e) = &res {
                    warn!("{}", e)
                };
                res.ok()
            })
            .collect();

        if let Err(e) = self.metrics_mailbox.send_and_forget(metric_readings) {
            warn!("Error adding metric sent to StatsD server: {}", e);
        }
    }
}

#[cfg(test)]
mod test {
    use crate::metrics::TakeMetrics;
    use insta::{assert_json_snapshot, with_settings};
    use rstest::{fixture, rstest};
    use ssf::ServiceMock;

    use super::*;

    #[rstest]
    #[case("test_counter:1|c", "test_gauge:2.0|g", "test_simple")]
    #[case("test-counter:1|c", "test-gauge:2.0|g", "test_simple_dashes")]
    #[case("test_counter:1|c", "test_counter:1|c", "test_counter_aggregation")]
    #[case(
        "test_counter:1|c\ntest_gauge:2.0|g",
        "test_counter:1|c\ntest_gauge:10.0|g",
        "test_counter_and_gauge_aggregation"
    )]
    #[case(
        "test_histo:100|h\ntest_another_histo:20.0|h",
        "test_one_more_histo:35|h\ntest_another_histo:1000.0|h",
        "test_histogram_aggregation"
    )]
    fn test_process_statsd_message(
        #[case] statsd_message_a: &str,
        #[case] statsd_message_b: &str,
        #[case] test_name: &str,
        mut fixture: Fixture,
    ) {
        // Process first StatsD test message
        fixture.server.process_statsd_message(statsd_message_a);

        // Process second StatsD test message
        fixture.server.process_statsd_message(statsd_message_b);
        with_settings!({sort_maps => true}, {
        assert_json_snapshot!(test_name, fixture.mock.take_metrics().unwrap());
        });
    }

    struct Fixture {
        server: StatsDServer,
        mock: ServiceMock<Vec<KeyedMetricReading>>,
    }
    #[fixture]
    fn fixture() -> Fixture {
        let mock = ServiceMock::new();
        let server = StatsDServer::new(mock.mbox.clone());

        Fixture { server, mock }
    }
}
