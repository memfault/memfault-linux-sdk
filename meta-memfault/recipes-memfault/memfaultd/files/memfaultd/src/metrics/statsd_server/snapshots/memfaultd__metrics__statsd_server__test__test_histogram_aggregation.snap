---
source: memfaultd/src/metrics/statsd_server/mod.rs
expression: fixture.mock.take_metrics().unwrap()
---
{
  "test_another_histo": 510.0,
  "test_histo": 100.0,
  "test_one_more_histo": 35.0
}
